src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs

```diff
@@ -809,31 +809,55 @@ private static LiftingResult UserDefinedBinaryOperatorCanBeLifted(TypeSymbol lef
             // SPEC: better than all other function members, then the function member invocation is ambiguous and a binding-time 
             // SPEC: error occurs.
 
-            // UNDONE: This is a naive quadratic algorithm; there is a linear algorithm that works. Consider using it.
+            int i;
             var candidates = result.Results;
-            for (int i = 1; i < candidates.Count; ++i)
+            for (i = 0; i < candidates.Count; i++)
             {
-                if (candidates[i].Kind != OperatorAnalysisResultKind.Applicable)
+                if (candidates[i].Kind == OperatorAnalysisResultKind.Applicable)
+                {
+                    break;
+                }
+            }
+
+            for (int j = i + 1; j < candidates.Count; j++)
+            {
+                if (candidates[j].Kind == OperatorAnalysisResultKind.Inapplicable)
                 {
                     continue;
                 }
 
-                // Is this applicable operator better than every other applicable method?
-                for (int j = 0; j < i; ++j)
+                var better = BetterOperator(candidates[i].Signature, candidates[j].Signature, left, right, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
                 {
-                    if (candidates[j].Kind == OperatorAnalysisResultKind.Inapplicable)
-                    {
-                        continue;
-                    }
-                    var better = BetterOperator(candidates[i].Signature, candidates[j].Signature, left, right, ref useSiteDiagnostics);
-                    if (better == BetterResult.Left)
-                    {
-                        candidates[j] = candidates[j].Worse();
-                    }
-                    else if (better == BetterResult.Right)
-                    {
-                        candidates[i] = candidates[i].Worse();
-                    }
+                    candidates[j] = candidates[j].Worse();
+                }
+                else if (better == BetterResult.Right)
+                {
+                    candidates[i] = candidates[i].Worse();
+                    i = j;
+                }
+            }
+
+            for (int j = 0; j < candidates.Count; j++)
+            {
+                if (j == i || candidates[j].Kind == OperatorAnalysisResultKind.Inapplicable)
+                {
+                    continue;
+                }
+
+                var better = BetterOperator(candidates[i].Signature, candidates[j].Signature, left, right, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
+                {
+                    candidates[j] = candidates[j].Worse();
+                }
+                else if (better == BetterResult.Right)
+                {
+                    candidates[i] = candidates[i].Worse();
+                    break;
+                }
+                else
+                {
+                    break;
```

Comment:
>What if a candidate[j] was previously marked as Worse by an earlier pass? Don't we need to mark this as Applicable?

When it was worse than something why should we mark it back as `Applicable`? The algorithm for methods, however, needs more detail about applicability of methods, which is provided in this PR.

>And don't we want to find all equally applicable members in the case that multiple exist?

I suppose we should continue the loop here just like the algorithm for methods, right?